cgen: if expr with index expr (when the element type of the array is struct)

https://github.com/vlang/v/issues/15664

Sorry, there's another situation that hasn't been tested:  When the element type of the array is struct.


**What did you do?**

```v
import rand

struct Foo {
	name string
}

fn main() {
	a := [Foo{'abc1'}, Foo{'abc2'}, Foo{'abc3'}]

	b := if true { a[rand.intn(a.len) or { 0 }].name } else { 'not found' }
	println(b)
}


```

**What did you expect to see?**

No error. 

**What did you see instead?**

```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/a.3808144713965031064.tmp.c:15291:36: error: expected ')'
        string b = (true ? (            _option_int _t1 = rand__intn(a.len);
                                                    ^
/tmp/v_501/a.3808144713965031064.tmp.c:15291:21: note: to match this '('
        string b = (true ? (            _option_int _t1 = rand__intn(a.len);
                           ^
/tmp/v_501/a.3808144713965031064.tmp.c:15291:59: error: expected expression
        string b = (true ? (            _option_int _t1 = rand__intn(a.len);
                                                                           ^
/tmp/v_501/a.3808144713965031064.tmp.c:15291:59: error: expected ':'
        string b = (true ? (            _option_int _t1 = rand__intn(a.len);
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

Shoves-MacBook:v_test_001 shove$ 
```

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -112,6 +112,9 @@ fn (mut g Gen) need_tmp_var_in_expr(expr ast.Expr) bool {
 				}
 			}
 		}
+		ast.SelectorExpr {
+			return g.need_tmp_var_in_expr(expr.expr)
+		}
 		else {}
 	}
 	return false

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -31,42 +31,88 @@ fn (mut g Gen) need_tmp_var_in_expr(expr ast.Expr) bool {
 	if is_noreturn_callexpr(expr) {
 		return true
 	}
-	if expr is ast.MatchExpr {
-		return true
-	}
-	if expr is ast.CallExpr {
-		if expr.is_method {
-			left_sym := g.table.sym(expr.receiver_type)
-			if left_sym.kind in [.array, .array_fixed, .map] {
+	match expr {
+		ast.IfExpr {
+			if g.need_tmp_var_in_if(expr) {
 				return true
 			}
 		}
-		if expr.or_block.kind != .absent {
+		ast.MatchExpr {
 			return true
 		}
-	}
-	if expr is ast.CastExpr {
-		return g.need_tmp_var_in_expr(expr.expr)
-	}
-	if expr is ast.ParExpr {
-		return g.need_tmp_var_in_expr(expr.expr)
-	}
-	if expr is ast.ConcatExpr {
-		for val in expr.vals {
-			if val is ast.CallExpr {
-				if val.return_type.has_flag(.optional) {
+		ast.CallExpr {
+			if expr.is_method {
+				left_sym := g.table.sym(expr.receiver_type)
+				if left_sym.kind in [.array, .array_fixed, .map] {
+					return true
+				}
+			}
+			if expr.or_block.kind != .absent {
+				return true
+			}
+		}
+		ast.CastExpr {
+			return g.need_tmp_var_in_expr(expr.expr)
+		}
+		ast.ParExpr {
+			return g.need_tmp_var_in_expr(expr.expr)
+		}
+		ast.ConcatExpr {
+			for val in expr.vals {
+				if val is ast.CallExpr {
+					if val.return_type.has_flag(.optional) {
+						return true
+					}
+				}
+			}
+		}
+		ast.IndexExpr {
+			if expr.or_expr.kind != .absent {
+				return true
+			}
+			if g.need_tmp_var_in_expr(expr.index) {
+				return true
+			}
+		}
+		ast.ArrayInit {
+			if g.need_tmp_var_in_expr(expr.len_expr) {
+				return true
+			}
+			if g.need_tmp_var_in_expr(expr.cap_expr) {
+				return true
+			}
+			if g.need_tmp_var_in_expr(expr.default_expr) {
+				return true
+			}
+			for elem_expr in expr.exprs {
+				if g.need_tmp_var_in_expr(elem_expr) {
 					return true
 				}
 			}
 		}
-	}
-	if expr is ast.IndexExpr {
-		if expr.or_expr.kind != .absent {
-			return true
+		ast.MapInit {
+			for key in expr.keys {
+				if g.need_tmp_var_in_expr(key) {
+					return true
+				}
+			}
+			for val in expr.vals {
+				if g.need_tmp_var_in_expr(val) {
+					return true
+				}
+			}
 		}
-		if g.need_tmp_var_in_expr(expr.index) {
-			return true
+		ast.StructInit {
+			if g.need_tmp_var_in_expr(expr.update_expr) {
+				return true
+			}
+			for field in expr.fields {
+				if g.need_tmp_var_in_expr(field.expr) {
+					return true
+				}
+			}
 		}
+		else {}
 	}
 	return false
 }

--- a/vlib/v/gen/c/match.v
+++ b/vlib/v/gen/c/match.v
@@ -28,9 +28,7 @@ fn (mut g Gen) need_tmp_var_in_match(node ast.MatchExpr) bool {
 			if branch.stmts.len == 1 {
 				if branch.stmts[0] is ast.ExprStmt {
 					stmt := branch.stmts[0] as ast.ExprStmt
-					if stmt.expr in [ast.CallExpr, ast.IfExpr, ast.MatchExpr, ast.StructInit]
-						|| (stmt.expr is ast.IndexExpr
-						&& (stmt.expr as ast.IndexExpr).or_expr.kind != .absent) {
+					if g.need_tmp_var_in_expr(stmt.expr) {
 						return true
 					}
 				}

--- a/vlib/v/tests/if_expr_with_index_expr_test.v
+++ b/vlib/v/tests/if_expr_with_index_expr_test.v
@@ -7,3 +7,15 @@ fn test_if_expr_with_index_expr() {
 	println(b)
 	assert true
 }
+
+struct Foo {
+	name string
+}
+
+fn test_if_expr_with_selector_expr() {
+	a := [Foo{'abc1'}, Foo{'abc2'}, Foo{'abc3'}]
+
+	b := if true { a[rand.intn(a.len) or { 0 }].name } else { 'not found' }
+	println(b)
+	assert true
+}

